### PR TITLE
Update: Add `computed` option for `quote-props`' as-needed

### LINKED ISSUE #5402: Proposing a "computed-property-as-needed" rule
With ES6 Computed Properties, one can do the following:

```javascript
let prop = 'b';
return {
  'a': 1,
  [prop]: 2,
  c: 3,
  ['d']: 4
}
```

Note how the brackets around the `['d']` property are superfluous. It could instead be written as `'d'` or `d`.

**When does this rule warn? Please describe and show example code:**

If enabled, the following would trigger:

```javascript
{
  ['x']: 1
}
```

Any property which does not need to be computed could trigger the rule.

**Is this rule preventing an error or is it stylistic?**

Stylistic

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

This rule affects basic ES6 syntax for situations where there are two equivalent ways to write the same code.

**Are you willing to create the rule yourself?**

I will attempt to if nobody else is interested.

### PR DESCRIPTION
**Pull request by BYK**
Fixes #5402.

### CODE REVIEW

File: docs/rules/quote-props.md [hunk 1]
```diff
@@ -216,9 +216,35 @@ and the following patterns _stop_ being problems:
...
 }
 ```
 
+When `computed` is set to `true`, the following patterns become problems:
+
+```js
+/*eslint quote-props: [2, "as-needed", { "computed": true }]*/
+
+var x = {
+    [100]: 1
```
Comment (BYK) [inline docs/rules/quote-props.md]: May be good to mention the interaction of this with the `numbers` option (this becomes a problem only if `numbers` is set to `false`)

File: docs/rules/quote-props.md [hunk 2]
```diff
@@ -216,9 +216,35 @@ and the following patterns _stop_ being problems:
 ```js
 var x = {
     "100": 1
+};
+```
+
+A `computed` flag, with default value `false`, can also be used as a modifier for the `"as-needed"` mode. When it is set to `true`, unnecessary property name computations will be reported.
```
Comment (nzakas) [inline docs/rules/quote-props.md]: Hmm, not sure `computed` is clear enough, as it's quite different from `numbers`, which require numeric properties to be quoted. I'd expect `computed` to then mean that computed properties should be quoted...but that's not what it does.

Maybe `allowComputed` with a default of `true`, which means don't check computed properties at all.

Eh, you know, I'm starting to think I was wrong about bundling this functionality with `quote-props`. :(

Comment (BYK) [inline docs/rules/quote-props.md]: >Eh, you know, I'm starting to think I was wrong about bundling this functionality with quote-props. :(

In all honesty, I felt the same when I was implementing this rule and writing the docs. I still feel like these should probably be in close proximity but may be a name change for the rule might be better?

As a stop-gap, I think `allowComputed` defaulting to true is also good. That said I think that's objectively clearer than `computed`.

Thinking again, `"as-needed", { computed: true }` actually makes sense. `computed` is confusing without that `as-needed` context.

**Comment by nzakas** [comment]
I believe we agreed that this wasn't the right approach, so closing.

**Pull request closed without merging**